### ISSUE TITLE
Logic error in `getThumbPath()`

### ISSUE THREAD

**Issue by khalwat** [craftcms/cms#2398]
The `getThumbUrl()` method in `craft\services\Assets` does this:

```php
public function getThumbUrl(Asset $asset, int $width, int $height = null, bool $generate = false): string
{
    if ($height === null) {
        $height = $width;
    }

    // Maybe a plugin wants to do something here
    // todo: remove the `size` key in 4.0
    $event = new GetAssetThumbUrlEvent([
        'asset' => $asset,
        'width' => $width,
        'height' => $height,
        'size' => max($width, $height),
        'generate' => $generate,
    ]);
    $this->trigger(self::EVENT_GET_ASSET_THUMB_URL, $event);

    // If a plugin set the url, we'll just use that.
    if ($event->url !== null) {
        return $event->url;
    }
```
So this is great, it gives plugins a chance to return a thumbnail. However, if `$event->generate` is `false`, and the thumbnail doesn't already exist, we should just be returning `null`. Everything's good so far.

Then it does:

```php
$path = $this->getThumbPath($asset, $width, $height, $generate);

if ($path === false) {
    return UrlHelper::actionUrl('assets/generate-thumb', [
        'uid' => $asset->uid,
        'width' => $width,
        'height' => $height,
        'v' => $asset->dateModified->getTimestamp(),
    ]);
}
```
Basically, get a path to the thumbnail, and only if it returns `false` should we return the AssetsController URL to generate the thumbnail.

I think the basic idea here is not to block on the creation of the thumbnails, but rather try to get them if they exist, and if they don't, spin off a request to the controller to actually generate them. Which makes sense.

However, in the `getThumbPath()` method, it does this:

```php
public function getThumbPath(Asset $asset, int $width, int $height = null, bool $generate = true)
{
    $ext = $asset->getExtension();

    // If it's not an image, return a generic file extension icon
    if (!Image::canManipulateAsImage($ext)) {
        $path = Craft::$app->getPath()->getAssetsIconsPath().DIRECTORY_SEPARATOR.strtolower($ext).'.svg';

        if (!file_exists($path)) {
            $svg = file_get_contents(Craft::getAlias('@app/icons/file.svg'));
            $extLength = strlen($ext);
            if ($extLength <= 3) {
                $textSize = '26';
            } else if ($extLength === 4) {
                $textSize = '22';
            } else {
                if ($extLength > 5) {
                    $ext = substr($ext, 0, 4).'…';
                }
                $textSize = '18';
            }
            $textNode = "<text x=\"50\" y=\"73\" text-anchor=\"middle\" font-family=\"sans-serif\" fill=\"#8F98A3\" font-size=\"{$textSize}\">".strtoupper($ext).'</text>';
            $svg = str_replace('<!-- EXT -->', $textNode, $svg);
            FileHelper::writeToFile($path, $svg);
        }

        return $path;
    }
```

In other words, if Craft thinks it can't be manipulated as an image, it's never going to return `false`, and instead will return the default icon string.

This means that many things, such as videos, that Craft can't manipulate as an image, plugins are never given a chance to generate a thumbnail for.

**Comment by khalwat** [comment]
So the formats that Craft can manipulate depends on the driver, but here's what Imagick returns:

http://phpimagick.com/Imagick/queryFormats

So because it can support `.mp4`, any `.mp4` videos can be manipulated, and thus my plugin can generate thumbnails for them.

However, video formats like `.webm` are not supported by Imagick, and thus even though `ffmpeg` (and my plugin) can generate a thumbnail for them, it's never given a chance to, due to the issue described in the previous post.

**Comment by khalwat** [comment]
Also because of:
```
    /**
     * Returns whether an image extension is considered manipulatable.
     *
     * @param string $extension
     *
     * @return bool
     */
    public static function canManipulateAsImage(string $extension): bool
    {
        $formats = Craft::$app->getImages()->getSupportedImageFormats();

        $alwaysManipulatable = ['svg'];
        $neverManipulatable = ['pdf'];

        $formats = array_merge($formats, $alwaysManipulatable);
        $formats = array_diff($formats, $neverManipulatable);

        return in_array(strtolower($extension), $formats);
    }
```
If some enterprising soul like @rog wanted to make thumbnail images from a PDF for the AdminCP (which would be awesome btw), he can't. Unless he wants to ignore `$event->generate` and block the rendering of the Assets index while the thumbnails are generated for the first time.

**Comment by andris-sevcenko** [comment]
@khalwat 

Looking at the current situation it definitely is not optimal. Seems like at some point we made the whole thumbnail generation process a little less plugin-friendly.

> so that the plugin can return false in the $event->url to indicate that the thumbnail isn't there, but it can generate one, similar to how $path works.

IMO this sounds dangerously obscure. I'd rather have a new property on the `GetAssetThumbUrlEvent` event work to indicate that a plugin is installed that would like to generate this thumbnail. Then, when coming through again only with `$generate` set to `true`, the plugin would return the URL. Basically, exactly what you described, only with a less obscure mechanic at work.

As a side note - as far as `canManipulateAsImage` is concerned, I'm not sure it'll change in the near future as there are parts where Craft itself does stuff with images where plugins are not involved, so taking plugin input at that point can have some unforeseen consequences.

**Comment by khalwat** [comment]
I only suggested the fix I did because it mirrors how the existing code works. I'd be more than happy to have something cleaner :)

**Comment by brandonkelly** [comment]
@khalwat Since you have your own thumbnail generation logic, why not just set `GetAssetThumbUrlEvent::url` to your own thumbnail generation URL, if `generate` is `true`?

**Comment by khalwat** [comment]
That's what I do @brandonkelly -- but the problem is the first time through `generate` is always `false`. It only ever is set to `true` if it's called from the `AssetController`... which never gets called because of the aforementioned logic error.

**Comment by brandonkelly** [comment]
@khalwat sorry re-read my comment, there was a crucial typo.

**Comment by khalwat** [comment]
Ah, okay. That's actually initially what I did... and it works.

The problem is that it blocks the Asset index view until _all_ of the thumbnails are generated if you do it this way. Which could potentially be awful.

This is I assume why it works the way it does for image thumbnails, it'll display them if they are already there, and if not it'll display a default icon. But then it'll kick off an AJAX request to generate them via the `AssetsController`, and display them asynchronously as they generate.

It's that last crucial step that's missing from the equation for thumbnails from file types that Craft doesn't think thumbnails can be generated from (that `canManipulateAsImage()` returns `false` on).

**Comment by khalwat** [comment]
(Incidentally, this same  problem is exhibited in the image previews that you just recently added to the Asset view... no preview image gets generated for file types that Craft doesn't think thumbnails can be generated from... even if a plugin is ready, willing, and able to generate them).

**Comment by brandonkelly** [comment]
To be clear, you have two options:

1. You can completely ignore `generate == false`, and just always generate the thumb if it doesn't exist yet. Which sounds like what you're currently doing, which has the caveat of slowing down the CP if there are a bunch of thumbnails that need to be generated.

2. You can create your own thumbnail generation controller action (similar to `assets/generate-thumb`), and then from the event handler, if you don't have a thumbnail yet AND `generate == false`, then return an action URL that points to your custom generation controller action.

In either case, you would always set `url` to _something_; never let it stay `null` if it’s for an asset you want to have thumbnail control over.

**Comment by khalwat** [comment]
In this case, we return `null` because that causes Craft generate its own thumbnail, either by returning a default thumbnail image, or if Craft deems the file format something it can generate a preview from, via the AJAX call it kicks off.

The issue is that Craft won't kick off that AJAX call for types it doesn't think it can handle (even if a plugin can).

But sure, if you'd prefer each plugin that wants to handle custom image previews does it via its own controller, I'll go that route.

**Comment by brandonkelly** [comment]
Yeah idk, it’s a possible solution. I guess ideally you don’t have to do that though. Another thing is, your plugin probably should also be able to hook into `getThumbPath()` in case something wants to get the thumbnail that way.

**Comment by brandonkelly** [comment]
Alright… just deprecated that `getAssetThumbUrl` event, and added a new event, `getThumbPath`. I would recommend that you use the new path event _instead_ of the deprecated URL event, setting its `path` property to either a generated thumbnail path on the server, or `false` if you want to be the one to handle it, but it doesn’t exist yet and `generate == false`. (And just let Craft worry about coming up with the URL for the thumbnail, based on its path.)

Let me know if that will work for you.

**Comment by khalwat** [comment]
Damn, you're good (and fast).

This should be perfect, will give it a whirl when RC9 ships out.

**Comment by khalwat** [comment]
So this works wonderfully. The first video is an `.webm` that neither Imagick nor Craft can do anything with; the other two are `.mp4` files that Imagick (and hence Craft) say they can manipulate:

<img width="714" alt="google chromescreensnapz058" src="https://user-images.githubusercontent.com/7570798/35893915-31dc213a-0b7e-11e8-9b91-a91816f779ba.png">

So thank you for that.

But (yeah, there's always a "but")... and you're probably sick of this, but I'm just noting it here for completeness.... but images that Craft can't manipulate, the new Asset preview view doesn't try even try to generate/display a thumbnail for them:

<img width="715" alt="google chromescreensnapz057" src="https://user-images.githubusercontent.com/7570798/35893961-6b381e48-0b7e-11e8-8fbe-2a770556afa3.png">

I'm also noting it because it will affect @michaelrog as well with the PDF image previews. Not a big deal, but perhaps there's an easy fix.

In an ideal world, displaying the Asset preview isn't dependent on whether it can be edited in the imager editor or not; rather the presence of the `Edit` button would depend on that.

**Comment by khalwat** [comment]
Yeah so as a quick test, I changed `getEditorHtml()` in `craft\elements\Asset` to look like this:

```
    /**
     * @inheritdoc
     */
    public function getEditorHtml(): string
    {
        if (!$this->fieldLayoutId) {
            $this->fieldLayoutId = Craft::$app->getRequest()->getBodyParam('defaultFieldLayoutId');
        }

        $html = '';

        $thumbSizes = [
            [380, 190],
            [760, 380],
        ];
        $assetsService = Craft::$app->getAssets();
        if ($assetsService->getThumbPath($this, $thumbSizes[0][0], $thumbSizes[0][1], false) !== null) {
            // Are they allowed to edit the image?
            $user = Craft::$app->getUser();
            $editable = (
                $user->checkPermission('deleteFilesAndFoldersInVolume:'.$this->volumeId) &&
                $user->checkPermission('saveAssetInVolume:'.$this->volumeId) &&
                $this->getSupportsImageEditor()
            );
```

...and it works as I'd expect, the thumbnail shows up:

<img width="714" alt="google chromescreensnapz060" src="https://user-images.githubusercontent.com/7570798/35895467-8cf7e304-0b85-11e8-8f0a-983dd2d8e73f.png">

...but no editor button, which is perfect.

**Comment by brandonkelly** [comment]
@khalwat Yep good point. Fixed now (49b9aafe).
